### PR TITLE
Avoid searching entire trees when looking for implicit constructors in find-refs

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
The current approach walks the whole tree for any file that we know has an implicit-object-creation expression in it (e.g. `new(...)`).  This is because we don't know what that expression will bind to, and we have to test each out.

However, walkign the entire tree looking for these is unnecessary.  We can instead search the docs containing such nodes for the `new` tokens in them, and see if they start a `new(...)` node.  This means we only have to walk those small subset of tokens.  This addresses about 5% of the cost when doing a FAR on a named-type:

image

### CODE REVIEW

File: src/Workspaces/Core/Portable/FindSymbols/FindReferences/FindReferenceCache.cs [hunk 1]
```diff
@@ -130,9 +131,19 @@ public SymbolInfo GetSymbolInfo(SyntaxNode node, CancellationToken cancellationT
...
         // otherwise, we can use the text of the document to quickly find candidates and test those directly.
-        return this.SyntaxTreeIndex.ProbablyContainsEscapedIdentifier(identifier)
-            ? _identifierCache.GetOrAdd(identifier, identifier => FindMatchingIdentifierTokensFromTree(identifier, cancellationToken))
-            : _identifierCache.GetOrAdd(identifier, _ => FindMatchingIdentifierTokensFromText(identifier, cancellationToken));
+        if (this.SyntaxTreeIndex.ProbablyContainsEscapedIdentifier(identifier))
+        {
+            return _identifierCache.GetOrAdd(
+                identifier,
+                identifier => FindMatchingIdentifierTokensFromTree(identifier, cancellationToken));
+        }
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/FindSymbols/FindReferences/FindReferenceCache.cs:139]: this logic is the smae.

File: src/Workspaces/Core/Portable/FindSymbols/FindReferences/FindReferenceCache.cs [hunk 2]
```diff
@@ -130,9 +131,19 @@ public SymbolInfo GetSymbolInfo(SyntaxNode node, CancellationToken cancellationT
...
+
+        return _identifierCache.GetOrAdd(
+            identifier,
+            identifier => FindMatchingTokensFromText(
+                identifier,
+                static (identifier, token, @this) => @this.IsMatch(identifier, token),
+                this, cancellationToken));
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/FindSymbols/FindReferences/FindReferenceCache.cs:146]: this is almost hte same.  just the addition of hte lambdaA+args to call nito a unified helper here.

File: src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/ConstructorSymbolReferenceFinder.cs [hunk 3]
```diff
@@ -190,15 +190,26 @@ private static bool IsPotentialReference(PredefinedType predefinedType, ISyntaxF
...
         IMethodSymbol symbol,
         FindReferencesDocumentState state,
         Action<FinderLocation, TData> processResult,
         TData processResultData,
         CancellationToken cancellationToken)
     {
+        if (!state.Cache.SyntaxTreeIndex.ContainsImplicitObjectCreation)
+            return;
+
+        // Note: only C# supports implicit object creation.  So we don't need to do anything special around case
```
Comment (ToddGrun) [inline src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/ConstructorSymbolReferenceFinder.cs:203]: >So we don't need to do anything special around case [](http://example.com/codeflow?start=61&length=51)

dumb question. If we don't need to worry about VB here, and this is the only caller of GetNewKeywordTokens, why does that code need to worry about VB?

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/ConstructorSymbolReferenceFinder.cs:203]: it technically doesn't.  it was just easier this way.

File: src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/ConstructorSymbolReferenceFinder.cs [hunk 4]
```diff
@@ -208,16 +219,12 @@ private static bool IsPotentialReference(PredefinedType predefinedType, ISyntaxF
...
-        void CollectMatchingReferences(
-            SyntaxNode node, FindReferencesDocumentState state, Action<FinderLocation, TData> processResult, TData processResultData)
+        foreach (var newKeywordToken in newKeywordTokens)
         {
-            if (node.RawKind != implicitObjectKind)
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var node = newKeywordToken.Parent;
+            if (node is null || node.RawKind != implicitObjectKind)
                 return;
```
Comment (ToddGrun) [inline src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/ConstructorSymbolReferenceFinder.cs:228]: >return; [](http://example.com/codeflow?start=16&length=7)

is returning still correct now that this is no longer in it's own method?

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/ConstructorSymbolReferenceFinder.cs:228]: nope!

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/ConstructorSymbolReferenceFinder.cs:228]: going to add test that demonstrates this.

**Review by ToddGrun — APPROVED** [review]
:shipit:

**Pull request merged**